subsys/bluetooth/host/smp.c

```diff
@@ -2504,10 +2510,24 @@ static uint8_t legacy_request_tk(struct bt_smp *smp)
 
 		break;
 	case PASSKEY_DISPLAY:
-		if (IS_ENABLED(CONFIG_BT_FIXED_PASSKEY) &&
-		    fixed_passkey != BT_PASSKEY_INVALID) {
+		if (IS_ENABLED(CONFIG_BT_FIXED_PASSKEY) && fixed_passkey != BT_PASSKEY_RAND) {
 			passkey = fixed_passkey;
-		} else  {
+		}
+
+#if defined(CONFIG_BT_APP_PASSKEY)
+		if (smp_auth_cb && smp_auth_cb->app_passkey) {
+			passkey = smp_auth_cb->app_passkey(conn);
+
+			if (passkey != BT_PASSKEY_RAND) {
+				if (passkey > 999999) {
+					return BT_SMP_ERR_INVALID_PARAMS;
```

Comment:
This should probably be `BT_SMP_ERR_UNSPECIFIED` because `BT_SMP_ERR_INVALID_PARAMS` means `The Invalid Parameters error code indicates that the command
length is invalid or that a parameter is outside of the specified range` and belongs to SMP. The error code returned here will be sent to peer device and peer may treat it incorrectly.

I think it worth to add a warning here so that application knows what happened.